_posts/2020-09-19-github-actions-with-faasd.md

```diff
@@ -0,0 +1,256 @@
+
+```sh
+export IP="143.110.184.36"
+export PASSWORD=$(ssh root@$IP "sudo cat /var/lib/faasd/secrets/basic-auth-password")
+```
+
+Assuming that you have faas-cli already installed, login to faasd with faas-cli with the following commands
+
+```sh
+export OPENFAAS_URL=http://$IP:8080
+echo $PASSWORD | faas-cli login --password-stdin -u admin -g $OPENFAAS_URL
+```
+
+## Create an OpenFaaS function
+
+Now that we've logged into faasd with faas-cli, we're ready to run our functions.
+I'm going to deploy a golang function for this tutorial.
+
+```sh
+# create a new directory for our project
+mkdir faasd-hello
+# pull golang-http templates
+faas-cli template store pull golang-http
+# create a new function with faas-cli
+faas-cli new faasd-hello --lang golang-http --prefix utsavanand2
+```
+
+If you look into the root directory of the project, faas-cli has created two additional files:
+
+```
+./faasd-hello/handler.go
+./faasd-hello.yml
+```
+
+The content of our handler looks something like this:
+![./faasd-hello/handler.go](/images/2020-09-19-github-actions-faasd/handler-go.jpg)
+
+The YAML file is used to configure the deployment and the build and runtime environment of the function
+~[./faasd-hello.yml](/images/2020-09-19-github-actions-faasd/faasd-hello-yaml.jpg)
+
+## Setup a workflow file for Github Actions
+
+In the root directory of the project run:
+
+```sh
+mkdir -p .github/workflows
+nano ./github/workflows/main.yml
+```
+
+And paste in the following contents into main.yml file
+
+```yaml
+name: CI-CD
+
+on:
+  push:
+    branches: 
+      - master
+
+jobs:
+  func-build:
+    runs-on: ubuntu-latest
+    steps:
+      - 
+        name: Checkout
+        uses: actions/checkout@v2
+      -
+        name: Pull template
+        run: |
+          docker run --rm -v `pwd`:`pwd` -w `pwd` \
+          openfaas/faas-cli:latest-root \
+          template store pull golang-http
+      - 
+        name: Running faas-cli skrinkwrap build
+        run: |
+          docker run --rm -v `pwd`:`pwd` -w `pwd` \
+          openfaas/faas-cli:latest-root \
+          build -f faasd-hello.yml --shrinkwrap
+      -
+        name: Login to DockerHub
+        if: success()
+        uses: docker/login-action@v1
+        with:
+          username: ${{ secrets.DOCKER_USERNAME }}
+          password: ${{ secrets.DOCKER_PASSWORD }}
+      - 
+        name: Login to OpenFaaS Gateway
+        run: |
+          docker run --rm -v $HOME:/root -w $HOME \
+          openfaas/faas-cli:latest-root \
+          login \
+          -p ${{ secrets.OPENFAAS_GATEWAY_PASSWD }} \
+          -g ${{ secrets.OPENFAAS_GATEWAY }} \
+          && \
+          sudo chown $USER:$USER -R $HOME/.openfaas && \
+          ls -lah $HOME
+      - 
+        name: Build and Push OpenFaaS function
+        run: |
+          # Update rights of build dir from root to current user \
+          # Build and push multi-arch image \
+          sudo chown $USER:$USER -R build/ && \
```

Comment:
As a user I would be confused by the amount of times `sudo` is used. Reading the post, I don't understand why this is here, so let's explain it for the user after the workflow with a comment.

Explain why it's needed and what it does.

I'd also like to know if there is a way to avoid this as it makes the use of the Docker image a little messy. Can we have faas-cli login / build write to a custom folder, or can we perform the task in /tmp/?

CC @LucasRoesler for input on resolving the use chown/sudo

Reply:
Yes I'll add comments to explain why we need it. After experimenting with about a dozen of Github Action builds failing I realised that the `build` and `.openfaas` dir didn't have the owner set to user running the action in the workflow.